src/Components/Components/src/DynamicComponent.cs

```diff
@@ -0,0 +1,104 @@
+        [Parameter]
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
+        public Type Type { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets a dictionary of parameters to be passed to the component.
+        /// </summary>
+        // Note that this deliberately does *not* use CaptureUnmatchedValues. Reasons:
+        // [1] The primary scenario for DynamicComponent is where the call site doesn't
+        //     know which child component it's rendering, so it typically won't know what
+        //     set of parameters to pass either, hence the developer most likely wants to
+        //     pass a dictionary rather than having a fixed set of parameter names in markup.
+        // [2] If we did have CaptureUnmatchedValues here, then it would become a breaking
+        //     change to ever add more parameters to DynamicComponent itself in the future,
+        //     because they would shadow any coincidentally same-named ones on the target
+        //     component. This could lead to application bugs.
+        [Parameter]
+        public IDictionary<string, object>? Parameters { get; set; }
+
+        /// <inheritdoc />
+        public void Attach(RenderHandle renderHandle)
+        {
+            _renderHandle = renderHandle;
+        }
+
+        /// <inheritdoc />
+        public Task SetParametersAsync(ParameterView parameters)
+        {
+            // This manual parameter assignment logic will be marginally faster than calling
+            // ComponentProperties.SetProperties.
+            foreach (var entry in parameters)
+            {
+                if (entry.Name.Equals(nameof (Type), StringComparison.OrdinalIgnoreCase))
+                {
+                    Type = (Type)entry.Value;
+                }
+                else if (entry.Name.Equals(nameof(Parameters), StringComparison.OrdinalIgnoreCase))
+                {
+                    Parameters = (IDictionary<string, object>)entry.Value;
+                }
+                else
+                {
+                    throw new InvalidOperationException(
+                        $"{nameof(DynamicComponent)} does not accept a parameter with the name '{entry.Name}'. To pass parameters to the dynamically-rendered component, use the '{nameof(Parameters)}' parameter.");
+                }
+            }
+
+            _renderHandle.Render(_cachedRenderFragment);
+            return Task.CompletedTask;
+        }
+
+        void Render(RenderTreeBuilder builder)
+        {
+            if (Type is null)
+            {
+                throw new InvalidOperationException($"{nameof(DynamicComponent)} requires a non-null value for the parameter {nameof(Type)}.");
+            }
+
+            builder.OpenComponent(0, Type);
+
+            if (Parameters != null)
+            {
+                foreach (var entry in Parameters)
+                {
+                    builder.AddAttribute(1, entry.Key, entry.Value);
```

Comment:
I'm not super familiar with the diffing behavior, what does it mean for all attributes to have the same sequence number? Should we be doing something special to ensure that we enumerate the dictionary in some consistent order (since dictionary has no enumeration guarantees)?

Reply:
Good point to raise. However I don't think we want to rely on enforcing consistent enumeration order. Presumably that would mean something like sorting the key-value pairs, which would immediately be the most expensive thing going on here.

In practice, I'm assuming that the evaluation order will (at least mostly) be consistent on successive renders. If it is, the attribute diffing rules will produce identical and optimal behavior regardless of whether the sequence numbers are all equal or if they are increasing.

If the evaluation order sometimes changes significantly between renders, then the attribute diffing system will also handle this just fine, but will involve more operations internally as the non-matching attribute names will trigger it going into its "dictionary building" attribute diffing mode which isn't as fast as the pure forward path it takes if the order hasn't changed.

Reply:
> "dictionary building" attribute diffing mode 

magic. got it. But yeah, in the most common case, i.e using a `Dictionary<TKey, TValue>`, the observed behavior is that you enumerate based on enumeration. While it's good to know that we don't rely on this behavior, it's fairly likely we're going to hit the fast past.